runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java

```diff
@@ -0,0 +1,155 @@
+    }
+
+    /** Recover a failed write at the given global address,
+     * driving it to completion by invoking the recovery
+     * protocol.
+     *
+     * When this function returns the given globalAddress
+     * is guaranteed to contain a committed value.
+     *
+     * If there was no data previously written at the address,
+     * this function will throw a runtime exception. The
+     * recovery protocol should -only- be invoked if we
+     * previously were overwritten.
+     *
+     * @param globalAddress     The global address to drive
+     *                          the recovery protocol
+     *
+     */
+    protected void recover(long globalAddress) {
+        // In chain replication, we started writing from the head,
+        // and propagated down to the tail. To recover, we start
+        // reading from the head, which should have the data
+        // we are trying to recover
+        int numUnits = layout.getSegmentLength(globalAddress);
+        log.debug("Recover[{}]: read chain head {}/{}", globalAddress, 1, numUnits);
+        ILogData ld = CFUtils.getUninterruptibly(layout.getLogUnitClient(globalAddress, 0)
+                .read(globalAddress)).getReadSet().getOrDefault(globalAddress, null);
+        // If nothing was at the head, this is a bug and we
+        // should fail with a runtime exception, as there
+        // was nothing to recover - if the head was removed
+        // due to a reconfiguration, a network exception
+        // would have been thrown and the client should have
+        // retried it's operation (in this case of a write,
+        // it should have read to determine whether the
+        // write was successful or not.
+        if (ld == null) {
+            throw new RuntimeException();
+        }
+        // now we go down the chain and write, ignoring any overwrite exception we get.
+        for (int i = 1; i < numUnits; i++) {
+            log.debug("Recover[{}]: write chain {}/{}", layout, i + 1, numUnits);
+            // In chain replication, we write synchronously to every unit
+            // in the chain.
+            try {
+                CFUtils.getUninterruptibly(
+                        layout.getLogUnitClient(globalAddress, i)
+                                // TODO: Currently, this call won't work until the log unit client is refactored.
+                                .write(globalAddress, null, null, ld, null)
+                        , OverwriteException.class);
+                // We successfully recovered a write to this member of the chain
+                log.debug("Recover[{}]: recovered write at chain {}/{}", layout, i + 1, numUnits);
+            } catch (OverwriteException oe) {
+                // This member already had this data (in some cases, the write might have
+                // been committed to all members, so this is normal).
+                log.debug("Recover[{}]: overwritten at chain {}/{}", layout, i + 1, numUnits);
```

Comment:
I think the best way would be to do this at the logging unit. When it reports overwrite exception, it can report if the data was the same or not (maybe using the checksum it's already generating anyway.)